core/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -310,7 +391,63 @@ public final void withContentOrSourceParamParserOrNull(CheckedConsumer<XContentP
      * Get the content of the request or the contents of the {@code source} param. Prefer {@link #contentOrSourceParamParser()} or
      * {@link #withContentOrSourceParamParserOrNull(CheckedConsumer)} if you need a parser.
      */
-    public final BytesReference contentOrSourceParam() {
+    public final Tuple<XContentType, BytesReference> contentOrSourceParam() {
+        if (hasContent()) {
+            if (xContentType == null) {
+                throw new IllegalStateException("Content-Type must be provided");
+            }
+            return new Tuple<>(xContentType, content());
+        }
+        String source = param("source");
+        if (source != null) {
+            BytesArray bytes = new BytesArray(source);
+            String typeParam = param("source_type");
+            final XContentType xContentType;
+            if (typeParam != null) {
+                xContentType = parseContentType(Collections.singletonList(typeParam));
+            } else {
+                xContentType = XContentFactory.xContentType(bytes);
```

Comment:
I agree but this would have worked with yaml as well, so I think we need to still support that here?